src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -2006,3 +2059,531 @@ ShouldInitiateMetadataSync(bool *lockFailure)
+	}
+
+	InsertShardPlacementRow(shardId, placementId, shardState, shardLength, groupId);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * EnsureShardPlacementMetadataIsSane ensures if the input parameters for
+ * the shard placement metadata is sane.
+ */
+static void
+EnsureShardPlacementMetadataIsSane(Oid relationId, int64 shardId, int64 placementId,
+								   int32 shardState, int64 shardLength, int32 groupId)
+{
+	if (shardId <= INVALID_SHARD_ID)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Invalid shard id: %ld", shardId)));
+	}
+	else if (!ShardExists(shardId))
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Shard id does not exists: %ld", shardId)));
+	}
+	else if (LookupShardRelationFromCatalog(shardId, true) != relationId)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Shard id does not belong to relation \"%s\": %ld",
+							   get_rel_name(relationId), shardId)));
+	}
+
+	if (placementId <= INVALID_PLACEMENT_ID)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Shard placement has invalid placement id "
+							   "(%ld) for shard(%ld)", placementId, shardId)));
+	}
+
+	if (!(shardState == SHARD_STATE_ACTIVE ||
+		  shardState == SHARD_STATE_TO_DELETE))
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Invalid shard state: %d", shardState)));
+	}
+
+	bool nodeIsInMetadata = false;
+	WorkerNode *workerNode =
+		PrimaryNodeForGroup(groupId, &nodeIsInMetadata);
+	if (!workerNode)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Node with group id %d for shard placement "
+							   "%ld does not exist", groupId, shardId)));
+	}
+}
+
+
+/*
+ * ShouldSkipMetadataChecks returns true if the current user is allowed to
+ * make any
+ */
+static bool
+ShouldSkipMetadataChecks(void)
+{
+	if (strcmp(EnableManualMetadataChangesForUser, "") != 0)
+	{
+		/*
+		 * EnableManualMetadataChangesForUser is a GUC which
+		 * can be changed by a super user. We use this GUC as
+		 * a safety belt in case the current metadata checks are
+		 * too restrictive and the operator can allow users to skip
+		 * the checks.
+		 */
+
+		/*
+		 * Make sure that the user exists, and print it to prevent any
+		 * optimization skipping the get_role_oid call.
+		 */
+		bool missingOK = false;
+		Oid allowedUserId = get_role_oid(EnableManualMetadataChangesForUser, missingOK);
+		if (allowedUserId == GetUserId())
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * citus_internal_update_placement_metadata is an internal UDF to
+ * update a row in pg_dist_placement.
+ */
+Datum
+citus_internal_update_placement_metadata(PG_FUNCTION_ARGS)
+{
+	int64 shardId = PG_GETARG_INT64(0);
+	int32 sourceGroupId = PG_GETARG_INT32(1);
+	int32 targetGroupId = PG_GETARG_INT32(2);
+
+	ShardPlacement *placement = NULL;
```

Comment:
Should we lock the relation id here as well?

Reply:
I didn't lock because we only allow changing groupId via this GUC, and all the calls for the same placement are already serialized by ` (shardid, groupid)` unique index on pg_dist_placement.